## CONTENT LIST
1. Importing Libraries
2. Importing Products Data
3. Data Consistency Checks
4. Task Questions

## 01. Importing Libraries

In [2]:
#Import Libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Products Data

In [3]:
#Importing Products Data
df_prods = pd.read_csv(r'C:\Users\chery\Desktop\CareerFoundry\Immersion Data Analytics\Achievement 4\02-2022 Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [4]:
#Importing Wrangled Orders Data
df_ords = pd.read_csv(r'C:\Users\chery\Desktop\CareerFoundry\Immersion Data Analytics\Achievement 4\02-2022 Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled_march8.csv', index_col = False)

## 03. Data Consistency Checks

### Mixed-Type Data

In [5]:
#Basic Stats on Orders df
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [7]:
#Create a Dataframe
df_test = pd.DataFrame()

In [8]:
#Create a mixed type column
df_test['mix'] = ['a','b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
#Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


### Find Missing Values

In [12]:
#Check for Missing data

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#Create subset for Missing data
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
#View Subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Address Missing Values

In [15]:
#Seeing Number of Rows in df including missing values
df_prods.shape

(49693, 5)

In [16]:
#Create new df excluding missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
#Seeing Number of Rows in New df excluding missing values
df_prods_clean.shape

(49677, 5)

### Find Duplicates

In [18]:
#Create new subset with only duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
#View subset 
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Address Duplicates

In [20]:
#Seeing Number of Rows in df including duplicates
df_prods_clean.shape

(49677, 5)

In [21]:
#Create new subset without duplicate rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
#Seeing Number of Rows in df excluding duplicates
df_prods_clean_no_dups.shape

(49672, 5)

### Export df_prds_clean_no_dups

In [23]:
#Exporting df_prods_clean_no_dups Dataframe

path = r'C:\Users\chery\Desktop\CareerFoundry\Immersion Data Analytics\Achievement 4\02-2022 Instacart Basket Analysis'

In [24]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data', 'products_checked_march9.csv'))

# 04. Task Questions

### Question 2

In [26]:
#Anything odd about checked df_prods basic statistics?

df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

*The max for 'prices' certainly seems off. It is hard to believe a standard grocery store would have any item that costs 99,999

### Question 3

In [27]:
#Check for mixed-type data in df_ords dataframe

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) !=df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird])>0:
        print (col)

In [28]:
print (df_ords.dtypes)

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object


### Question 4

In [ ]:
#Fix mixed-type data in df_ords dataframe

*There does not appear to be any mixed-type data to fix

### Question 5

In [29]:
#Run check for missing values in df_ords dataframe

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

*The 206,209 missing values under 'days_since_prior_order' column most likely is new customers who have never placed an order before

### Question 6

In [38]:
#Address missing values - Create new column labeling 'New Customer' if missing value under 'days_since_prior_order' column. Label as 'Repeat Customer' for any value higher or equal to 0

def x(row):
 if row['days_since_prior_order'] >= 0:
    return 'Repeat Customer'
 else: return 'New Customer'

In [39]:
#Apply new function to df
df_ords['customer_status']= df_ords.apply(x,axis=1)

In [40]:
print(df_ords)

         Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  customer_status  
0                

To address the missing data I created a new column flagging any customers with '0' or greater for 'days_since_prior_order' as a 'Repeat Customer'. Any customer that had missing data for this column was labeled as 'New Customer'. It did not make sense to delete these rows because there are over 200,000 records and data on new customers can be very valuable. It's also not logical to replace the missing data with the median/mean because these first time customers will most likely return and their future orders will be added to the data & could possible influence anaylsis.

### Question 7

In [41]:
#Run check for duplicates in df_ords

df_ords_dups = df_ords[df_ords.duplicated()]

In [42]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, customer_status]
Index: []

### Question 8

In [43]:
#Address duplicates

I don't believe there are any duplicates to address as seen in the above output

### Question 9

In [47]:
#Export dfs (df_prods exported earlier in notebook)

df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data', 'orders_checked_march9.csv'))